test/tcp_echo_client/tcp_echo_client.cpp

```diff
@@ -20,17 +20,201 @@
+
+    /* Create the socket */
+    err = api->init();
+    if (!TEST_EQ(err, SOCKET_ERROR_NONE)) {
+        TEST_RETURN();
+    }
+    s.impl = NULL;
+    err = api->create(&s, SOCKET_AF_INET4, SOCKET_STREAM, &tcp_ec_client_cb);
+    if (!TEST_EQ(err, SOCKET_ERROR_NONE))
+    {
+        TEST_EXIT();
+    }
+    err = api->bind(&s, &local_saddr, local_port);
+    if (!TEST_EQ(err, SOCKET_ERROR_NONE)) {
+        TEST_RETURN();
+    }
+
+    err = api->str2addr(&s, &srv_saddr, srv_addr_s);
+    TEST_EQ(err, SOCKET_ERROR_NONE);
+    tcp_ec_sock_addr_port_dump("server", &srv_saddr, srv_port);
+
+
+    tcp_ec_timeout_g = 0;
+    tcp_ec_connected = 0;
+    to.attach(onTimeout, TCP_EC_SOCKET_TEST_TIMEOUT);
+    err = api->connect(&s, &srv_saddr, srv_port);
+    TEST_EQ(err, SOCKET_ERROR_NONE);
+    if (err!=SOCKET_ERROR_NONE) {
+        printf("err = %d\r\n", err);
+    }
+    while (!tcp_ec_connected && !tcp_ec_timeout_g)
+    {
+        __WFI();
+    }
+    to.detach();
+    TEST_EQ(tcp_ec_timeout_g, 0);
+
+    tcp_ec_client_tx_done_g = false;
+    tcp_ec_client_rx_done_g = false;
+    tcp_ec_timeout_g = 0;
+    to.attach(onTimeout, TCP_EC_SOCKET_TEST_TIMEOUT);
+    err = api->send(&s, shutdown_cmd, strlen(shutdown_cmd));
+    if (!TEST_EQ(err, SOCKET_ERROR_NONE)) {
+        TEST_PRINT("Failed to send shutdown command %u bytes. err=%s\r\n", strlen(shutdown_cmd), socket_strerror(err));
+    }
+    else
+    {
+        size_t tx_bytes = 0;
+        do {
+            /* Wait for the onSent callback */
+            while (!tcp_ec_timeout_g && !tcp_ec_client_tx_done_g) {
+                __WFI();
+            }
+            if (!TEST_EQ(tcp_ec_timeout_g,0)) {
+                break;
+            }
+            if (!TEST_NEQ(tcp_ec_client_tx_info_g.sentbytes, 0)) {
+                break;
+            }
+            tx_bytes += tcp_ec_client_tx_info_g.sentbytes;
+            if (tx_bytes < strlen(shutdown_cmd)) {
+                tcp_ec_client_tx_done_g = false;
+                continue;
+            }
+            to.detach();
+            TEST_EQ(tx_bytes, strlen(shutdown_cmd));
+            break;
+        } while (1);
+    }
+
+    /* this test depends on us closing the socket before the remote peer does
+     * if the remote server were to close first then the closing a closed connection
+     * would return an error. For this reason the sal_tcpserver waits 5s before terminating
+     * the server */
+    err = api->close(&s);
+    TEST_EQ(err, SOCKET_ERROR_NONE);
+
+    /* destroy the socket */
+    err = api->destroy(&s);
+    TEST_EQ(err, SOCKET_ERROR_NONE);
+
+test_exit:
+    TEST_RETURN();
+    return 0;
+}
+
+
+void app_start(int , char **)
 {
-    (void) argc;
-    (void) argv;
-    MBED_HOSTTEST_TIMEOUT(20);
-    MBED_HOSTTEST_SELECT(tcpecho_client_ext_auto);
-    MBED_HOSTTEST_DESCRIPTION(Socket Abstraction Layer TCP Echo Client test);
-    MBED_HOSTTEST_START("SAL_TCP_ECHO_CLIENT");
+    MBED_HOSTTEST_TIMEOUT(60);
```

Comment:
- agreed, this value selection this should be document
- its because the test can run slowly on Jenkins infrastructure where multiple jobs are competing for resources
- this value was found to work reliably on jenk_slave_1